### ISSUE TITLE
Issue with time taken by graph run()  

### ISSUE THREAD

**Issue by ved27** [ARM-software/ComputeLibrary#489]
Hi,

version : Computelibrary 18.05
Platform : arm7a
Operating System: linux

I'm running a slightly modified example/graph_mobilenet.cpp  with target = OPENCL

   void do_run()
    {

        for(int i = 0; i < 5; i++)
         {
             struct timeval start , end;       
         
          gettimeofday(&start, NULL);
        
             graph.run();            
         
            gettimeofday(&end, NULL);
         
             printf("%ld \n ",((end.tv_sec*1000000 + end.tv_usec) - (start.tv_sec*1000000 + start.tv_usec))); 
        }
    }
    

The result is: (in micro sec)
32108
452369
523478
425689
436898 
    
I wonder why the last 4 results are huge around 0.5 sec , while the first one is ~32ms 

Kindly help me with the possible reasons of this issue,
Please let me know if you need more details

Eagerly awaiting your reply,

**Comment by AnthonyARM** [comment]
32ms for Mobilenet sounds unrealistic, which platform is this on ? Do you pass any parameters to the examples ?

**Comment by ved27** [comment]
Hi,

Platform : ARM7a , using Mali G51 

scons Werror=1 -j8 debug=0 asserts=0 neon=1 opencl=1 openmp=1 cppthreads=1 embed_kernels=1 os=linux arch=armv7a

target : 1 , which is OpenCL 
My input size is 3x300x300 (C x H  x W) 

The modified mobilenet graph has nodes for convolution layers added additionally at the end of original graph example .

Please note that in the following code for profiling the runtime, I'm not considering CLScheduler::get().sync() .

void do_run()
{

    for(int i = 0; i < 5; i++)
     {
         struct timeval start , end;       
     
      gettimeofday(&start, NULL);
    
         graph.run();            
     
        gettimeofday(&end, NULL);
     
         printf("%ld \n ",((end.tv_sec*1000000 + end.tv_usec) - (start.tv_sec*1000000 + start.tv_usec))); 
    }
}


Kindly let me know if you need more info,

**Comment by gmiodice** [comment]
Hi @ved27,

do you pass any input image to your test? Moreover are you using G51 or G71 Mali GPU?

Thanks

**Comment by ved27** [comment]
Hi @gmiodice 

I'm reading 3 x 300 x 300 image from a  .txt file . I'm using Mali G51 GPU

**Comment by AnthonyARM** [comment]
If you haven't changed anything in the graph.run() and the accessors then there should be a blocking map at the end of run() which should have the same result as `sync()`

Looking at your times, it looks like you've removed the blocking call at the end of run() and therefore the first run will just enqueue jobs and not wait for the results, then the second run() will block at the beginning of run() and wait for the first one to complete before enqueuing its own jobs, etc.
Basically all your times are shifted by 1.

**Comment by ved27** [comment]
Hi @AnthonyARM 

You are right. I also tried using the blocking map at the end of each loop and results of this are similar to using sync() .

If possible can you please share the model and npy files you used to test the graph.

Can you please let me know your tested runtimes for examples/graph_mobilet.cpp ?
How long is each loop taking in your testing scenarios?

**Comment by ved27** [comment]
Hi ,

So, you too are getting the runtime of ~0.45 sec for examples/graph_mobilenet.cpp on Mali G51 ?
Is the figure realistic ?

**Comment by AnthonyARM** [comment]
we just use `examples/graph_mobilenet.cpp` for benchmarking. NPY files are just used for validation, for benchmarking it just slows things down because of the file accesses, and doesn't change anything to the actual execution time.
So a stock graph_mobilenet.cpp is the best thing to use for benchmarking
On a G71 MP8 @ 807MHz we get ~25ms
No idea what the performance of a G51 is supposed to be, it also depends on how many cores you've got

**Comment by ved27** [comment]
Hi @AnthonyARM 

In your previous reply you mentioned the following quote, 
""
32ms for Mobilenet sounds unrealistic, which platform is this on ? Do you pass any parameters to the examples ?
""
Can I please know why was it unrealistic ?

Also, as you mentioned "On a G71 MP8 @ 807MHz we get ~25ms",
Can I know the input size / Convolution method / math hint  and other parameters you configured ?

**Comment by ved27** [comment]
Hi,

As I checked , we have got Mali G51 - 2 Core and frequency is < 807MHz .
Could not know the consistent frequency of it during the run time .
Can you please give me some input to check the operating frequency when I run the application ?

May be ~0.3 - 0.4 sec with 2 cores and < 600MHz is okay ?

**Comment by ved27** [comment]
Hi ,

Thank you much for the info

The path for that might be different in my current setup . 
There is a DVFS setting which shows the value 100 (MHz) and we are unable to set it to max frequency. And this is 2 Core Mali G51 .

I get the similar ~0.3- 0.4 sec on CL , NEON as well . 
I would like to know if there is a performance comparison with your setup of G71 ,@807MHz MP8 vs that of G51 also how the runtime varies with parameters like frequency ,cores and version of ARM GPU 
 

Thanks again

**Comment by GeorgeARM** [comment]
Hello @ved27,

We haven't a G51 configuration at hand at the moment. As was mentioned above the 32ms in mobilenet is more that possible as we get less that 25ms in G71. The issue title is out of sync thus I am closing this issue. Create a new one if needed but as @AnthonyARM said performance depends on the number of cores, frequency etc.
